src/plugin.py

```diff
@@ -103,40 +122,72 @@ def test_stream(self: Corva, fn, event, *args, **kwargs):
 
             return func(self, fn, events, *args, **kwargs)
 
-        return test_stream
+        return _patch_corva
 
-    with mock.patch.object(Corva, 'stream', decorator(Corva.stream)):
+    with mock.patch.object(Corva, 'stream', patch_corva(Corva.stream)):
         yield
 
 
 @contextlib.contextmanager
 def patch_scheduled():
     """Patches scheduled runner."""
 
-    from corva.application import Corva
+    # imports are local to avoid loading packages, on the first plugin run
+    from corva.application import Corva, scheduled_runner
+
+    def patch_corva(func):
+        def _patch_corva(
+            self: Corva,
+            fn: Callable,
+            event: Union[dict, List[dict], List[List[dict]]],
+            *args,
+            **kwargs,
+        ):
+            """Automatically adds essential fields to event in Corva.scheduled."""
 
-    def decorator(func):
-        def test_scheduled(self: Corva, fn, event, *args, **kwargs):
             events = copy.deepcopy(event)
+
+            # cast event to expected type List[List[dict]], if needed
             if not isinstance(events, list):
+                # allow users to send event as dict
                 events = [events]
             if not isinstance(events[0], list):
+                # allow users to send event as List[dict]
                 events = [events]
 
             for i in range(len(events)):
                 for event in events[i]:
+                    # do not override the values, if provided by user
                     event.setdefault('app_connection', int())
                     event.setdefault('app_stream', int())
 
             return func(self, fn, events, *args, **kwargs)
 
-        return test_scheduled
+        return _patch_corva
+
+    def patch_runner(func):
+        def _patch_runner(fn, context):
+            # context.api is accessed two times in scheduled_runner:
+            #   1. To pass an Api instance to the user function.
+            #   2. To make a POST API call.
+            # Patch context.api to return Mock instance, when accessed the second time.
+            # This is needed for following reasons:
+            #   1. To avoid POSTing in user tests.
+            #   2. To avoid adding our POST mock to users Api instance.
+            type(context).api = mock.PropertyMock(
+                side_effect=[context.api, mock.Mock()]
+            )
```

Comment:
`type(context).api` looks like magic. Is there any other way to mock `api` property?

Reply:
In this case, I don't think there is an alternative. https://stackoverflow.com/a/54441868